main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -2540,6 +2540,41 @@ object Typer {
           elementTypes <- traverseM(elms)(visit)
         } yield Type.mkTuple(elementTypes, loc)
 
+
+      case KindedAst.Pattern.Record(pats, pat, tvar, loc) =>
+        val ps = traverseM(pats) {
+          case KindedAst.Pattern.Record.RecordFieldPattern(field, tpe, pat1, loc1) => (tpe, pat1) match {
+            case (Some(t), Some(p)) =>
+              // { x : Type = pattern ... }
+              val freshRowVar = Type.freshVar(Kind.RecordRow, loc1)
+              val recordRow = Type.mkRecordRowExtend(field, t, freshRowVar, loc1)
+              for {
+                t1 <- visit(p)
+                _ <- unifyTypeM(recordRow, t1, loc1)
+                resultTyp = t
+              } yield resultTyp
```

Comment:
@mlutze I think I might need some help with this.

I'm trying to come up with the type rules for the pattern `{ field : Type = Pattern }`. Note that there may be several of these, i.e.,
```ocaml
{ field1 : Type = Pattern, field2, field3 : Type }
```
hence the `...` in the comment in the code.

Is it correct to assume that the type of the field is a `recordRowExtend` type?
Furthermore, should it be unified with the constraints from the pattern (`t1 <- visit(p)`)?
Finally, what is the resulting type? I'm assuming that if all goes well, it should be the type of the provided type annotation? Or is it the `recordRowExtend` type?

Reply:
First I will confuse you by answering your questions.

> Is it correct to assume that the type of the field is a `recordRowExtend` type?

No. Each field has the kind `Type`.

> Furthermore, should it be unified with the constraints from the pattern (`t1 <- visit(p)`)?

Yes. In particular, you should use the `expectTypeM` function because we expect the pattern to match that type.

> Finally, what is the resulting type? I'm assuming that if all goes well, it should be the type of the provided type annotation? Or is it the `recordRowExtend` type?

Yes, in this case you can use the provided type annotation as the result type.

---- 

I think there is some confusion about the structure of record types.

These are the relevant type constructors and their kinds.
- `Record: RecordRow -> Type`
- `RecordRowExtend(label): Type -> RecordRow -> Type`
- `RecordRowEmpty: RecordRow`

A record is a wrapper around a record row. A record row is essentially a list-like structure, where `RecordRowExtend(label)` is `Cons` and `RecordRowEmpty` is `Nil`.

The record type `{x: String, y: Bool}` has the following structure:

```
Record[
  RecordRowExtend(x)[
    String,
    RecordRowExtend(y)[
      Bool,
      RecordRowEmpty
    ]
  ]
]
```

For your purposes here, I propose that while traversing your list of patterns, you do not pay attention to the record structure above. Just infer the type of each field. That means you don't need to deal with anything involving `RecordRow` inside your traversal.

After you have your list of fields with their inferred types, you can use them to build the final record type.

Reply:
Ah, okay, that helps a lot, thanks!

Reply:
Am I correct in assuming that handling correspondence of the matched type and the pattern is handled by the `match-case` expression? This would mean that I only have to infer the type of the pattern, and the `match-case` type rules will ensure that the record types actually match (with correct fields etc.)?

Reply:
That is correct.